tests/auth/ldap_test.py

```diff
@@ -6,33 +6,92 @@
 """
 import pytest
 
+# module must have ldap.INVALID_CREDENTIALS
+INVALID_CREDENTIALS = Exception
+
 
 def test_happy_path():
-    from sirepo import srunit
     from pykern.pkcollections import PKDict
     from pykern import pkinspect
+    from sirepo import srunit
     import sys
 
     sys.modules["ldap"] = pkinspect.this_module()
 
     with srunit.quest_start(cfg={"SIREPO_AUTH_METHODS": "ldap"}) as qcall:
         from pykern import pkunit
-        from pykern import pkdebug
+        from sirepo import auth_db
 
         r = qcall.call_api(
             "authLdapLogin",
             data=PKDict(
                 simulationType="myapp",
-                email="vagrant@xxxxradiasoft.net",
+                email="vagrant@radiasoft.net",
                 password="vagrant",
             ),
         )
         pkunit.pkeq(200, r.status_as_int())
         pkunit.pkre("location.*/complete-registration", r.content_as_str())
 
+        m = qcall.auth_db.model("AuthEmailUser")
+        u = m.unchecked_search_by(unverified_email="vagrant@radiasoft.net")
+        pkunit.pkne(None, u)
+
+
+def test_cred_validation():
+    from pykern.pkcollections import PKDict
+    from pykern import pkinspect
+    from sirepo import srunit
+    import sys
+
+    sys.modules["ldap"] = pkinspect.this_module()
+
+    with srunit.quest_start(cfg={"SIREPO_AUTH_METHODS": "ldap"}) as qcall:
+        from pykern import pkunit
+
+        r = qcall.call_api(
+            "authLdapLogin",
+            data=PKDict(
+                simulationType="myapp",
+                email="",
```

Comment:
Here I would test invalid password and have a valid email so that it goes all the way though to ldap.bind.